apriltag.c

```diff
@@ -802,19 +810,35 @@ static void refine_edges(apriltag_detector_t *td, image_u8_t *im_orig, struct qu
                 // gradient more precisely, but are more sensitive to
                 // noise.
                 double grange = 1;
-                int x1 = x0 + (n + grange)*nx;
-                int y1 = y0 + (n + grange)*ny;
-                if (x1 < 0 || x1 >= im_orig->width || y1 < 0 || y1 >= im_orig->height)
+                double x1 = x0 + (n + grange)*nx;
+                double y1 = y0 + (n + grange)*ny;
+                int x1i = floor(x1);
+                int y1i = floor(y1);
+                double a1 = x1 - x1i;
+                double b1 = y1 - y1i;
```

Comment:
Do you just want to get the integral and fractional part? If so, just use [`modf`](https://en.cppreference.com/w/c/numeric/math/modf).